ethkey/src/math.rs

```diff
@@ -0,0 +1,122 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::{SECP256K1, Public, Secret, Error};
+use secp256k1::key;
+
+/// Inplace multiply public key by secret key (EC point * scalar)
+pub fn public_mul_secret(public: &mut Public, secret: &Secret) -> Result<(), Error> {
+	let key_secret = key::SecretKey::from_slice(&SECP256K1, &secret[..])?;
+	let mut key_public = to_secp256k1_public(public)?;
+	key_public.mul_assign(&SECP256K1, &key_secret)?;
+	set_public(public, &key_public);
+	Ok(())
+}
+
+/// Inplace add one public key to another (EC point + EC point)
+pub fn public_add(public: &mut Public, other: &Public) -> Result<(), Error> {
+	let mut key_public = to_secp256k1_public(public)?;
+	let other_public = to_secp256k1_public(other)?;
+	key_public.add_assign(&SECP256K1, &other_public)?;
+	set_public(public, &key_public);
+	Ok(())
+}
+
+/// Inplace add one secret key to another (scalar + scalar)
+pub fn secret_add(secret: &mut Secret, other: &Secret) -> Result<(), Error> {
+	let mut key_secret = to_secp256k1_secret(secret)?;
+	let other_secret = to_secp256k1_secret(other)?;
+	key_secret.add_assign(&SECP256K1, &other_secret)?;
+
+	*secret = key_secret.into();
+	Ok(())
+}
+
+/// Inplace subtract one secret key from another (scalar - scalar)
+pub fn secret_sub(secret: &mut Secret, other: &Secret) -> Result<(), Error> {
+	let mut key_secret = to_secp256k1_secret(secret)?;
+	let mut other_secret = to_secp256k1_secret(other)?;
+	other_secret.mul_assign(&SECP256K1, &key::MINUS_ONE_KEY)?;
+	key_secret.add_assign(&SECP256K1, &other_secret)?;
+
+	*secret = key_secret.into();
+	Ok(())
+}
+
+/// Inplace multiply one secret key to another (scalar * scalar)
+pub fn secret_mul(secret: &mut Secret, other: &Secret) -> Result<(), Error> {
+	let mut key_secret = to_secp256k1_secret(secret)?;
+	let other_secret = to_secp256k1_secret(other)?;
+	key_secret.mul_assign(&SECP256K1, &other_secret)?;
+
+	*secret = key_secret.into();
+	Ok(())
+}
+
+/// Inplace inverse secret key (1 / scalar)
+pub fn secret_inv(secret: &mut Secret) -> Result<(), Error> {
+	let mut key_secret = to_secp256k1_secret(secret)?;
+	key_secret.inv_assign(&SECP256K1)?;
+
+	*secret = key_secret.into();
+	Ok(())
+}
+
+/// Compute power of secret key inplace (secret ^ pow).
+/// This function is not intended to be used with large powers.
+pub fn secret_pow(secret: &mut Secret, pow: usize) -> Result<(), Error> {
+	if pow == 0 {
+		*secret = key::ONE_KEY.into();
+		return Ok(());
+	}
+	if pow == 1 {
+		return Ok(());
+	}
+
+	let secret_copy = secret.clone();
+	for _ in 1..pow {
+		secret_mul(secret, &secret_copy)?;
+	}
+
+	Ok(())
+}
+
+/// Return base point of secp256k1
+pub fn generation_point() -> Public {
```

Comment:
isn't it the same as taking GENERATOR_X, GENERATOR_Y from here using less cpu cycles?

https://github.com/ethcore/rust-secp256k1/blob/master/src/constants.rs#L53